File: todo/api-php/endpoints/tasks.php
<?php
/**
 * Endpoints de Tarefas
 * Todo App - API PHP
 */

function handleTasks(?string $id, string $method): void {
    $auth = requireAuth();
    
    switch ($method) {
        case 'GET':
            if ($id) {
                getTask($auth['user_id'], $id);
            } else {
                getAllTasks($auth['user_id']);
            }
            break;
        case 'POST':
            createTask($auth['user_id']);
            break;
        case 'PUT':
            if ($id) updateTask($auth['user_id'], $id);
            break;
        case 'DELETE':
            if ($id) deleteTask($auth['user_id'], $id);
            break;
        default:
            jsonResponse(false, 'Método não permitido', null, 405);
    }
}

function getAllTasks(string $userId): void {
    $pdo = getDB();
    
    $stmt = $pdo->prepare("
        SELECT t.id, t.list_id, t.title, t.description, t.completed, t.reminder, t.created_at, t.updated_at
        FROM tasks t
        JOIN todo_lists l ON t.list_id = l.id
        WHERE l.user_id = ? AND t.deleted_at IS NULL AND l.deleted_at IS NULL
        ORDER BY t.completed ASC, t.created_at DESC
    ");
    $stmt->execute([$userId]);
    $tasks = $stmt->fetchAll();
    
    // Converter completed para boolean
    foreach ($tasks as &$task) {
        $task['completed'] = (bool) $task['completed'];
    }
    
    jsonResponse(true, 'Tarefas carregadas', $tasks);
}

function getTasksByList(string $listId): void {
    $auth = requireAuth();
    $pdo = getDB();
    
    // Verificar propriedade da lista
    $stmt = $pdo->prepare("SELECT id FROM todo_lists WHERE id = ? AND user_id = ?");
    $stmt->execute([$listId, $auth['user_id']]);
    
    if (!$stmt->fetch()) {
        jsonResponse(false, 'Lista não encontrada', null, 404);
    }
    
    $stmt = $pdo->prepare("
        SELECT id, list_id, title, description, completed, reminder, created_at, updated_at
        FROM tasks 
        WHERE list_id = ? AND deleted_at IS NULL
        ORDER BY completed ASC, created_at DESC
    ");
    $stmt->execute([$listId]);
    $tasks = $stmt->fetchAll();
    
    foreach ($tasks as &$task) {
        $task['completed'] = (bool) $task['completed'];
    }
    
    jsonResponse(true, 'Tarefas carregadas', $tasks);
}

function getTask(string $userId, string $taskId): void {
    $pdo = getDB();
    
    $stmt = $pdo->prepare("
        SELECT t.id, t.list_id, t.title, t.description, t.completed, t.reminder, t.created_at, t.updated_at
        FROM tasks t
        JOIN todo_lists l ON t.list_id = l.id
        WHERE t.id = ? AND l.user_id = ? AND t.deleted_at IS NULL
    ");
    $stmt->execute([$taskId, $userId]);
    $task = $stmt->fetch();
    
    if (!$task) {
        jsonResponse(false, 'Tarefa não encontrada', null, 404);
    }
    
    $task['completed'] = (bool) $task['completed'];
    
    jsonResponse(true, 'Tarefa encontrada', $task);
}

function createTask(string $userId): void {
    $data = getJsonInput();
    
    $id = $data['id'] ?? generateUUID();
    $listId = $data['list_id'] ?? '';
    $title = trim($data['title'] ?? '');
    $description = trim($data['description'] ?? '');
    $completed = (bool) ($data['completed'] ?? false);
    $reminder = $data['reminder'] ?? null;
    $createdAt = $data['created_at'] ?? date('Y-m-d H:i:s');
    
    if (empty($listId) || empty($title)) {
        jsonResponse(false, 'Lista e título são obrigatórios', null, 400);
    }
    
    $pdo = getDB();
    
    // Verificar propriedade da lista
    $stmt = $pdo->prepare("SELECT id FROM todo_lists WHERE id = ? AND user_id = ? AND deleted_at IS NULL");
    $stmt->execute([$listId, $userId]);
    
    if (!$stmt->fetch()) {
        jsonResponse(false, 'Lista não encontrada', null, 404);
    }
    
    // Verificar se já existe (para sincronização)
    $stmt = $pdo->prepare("SELECT id FROM tasks WHERE id = ?");
    $stmt->execute([$id]);
    
    if ($stmt->fetch()) {
        // Atualizar se já existe
        $stmt = $pdo->prepare("
            UPDATE tasks 
            SET title = ?, description = ?, completed = ?, reminder = ?, updated_at = NOW(), deleted_at = NULL
            WHERE id = ?
        ");
        $stmt->execute([$title, $description, $completed ? 1 : 0, $reminder, $id]);
    } else {
        // Criar nova
        $stmt = $pdo->prepare("
            INSERT INTO tasks (id, list_id, title, description, completed, reminder, created_at) 
            VALUES (?, ?, ?, ?, ?, ?, ?)
        ");
        $stmt->execute([$id, $listId, $title, $description, $completed ? 1 : 0, $reminder, $createdAt]);
    }
    
    // Log de sincronização
    logSync($userId, 'task', $id, 'create');
    
    jsonResponse(true, 'Tarefa criada com sucesso', [
        'id' => $id,
        'list_id' => $listId,
        'title' => $title,
        'description' => $description,
        'completed' => $completed,
        'reminder' => $reminder,
        'created_at' => $createdAt
    ], 201);
}

function updateTask(string $userId, string $taskId): void {
    $data = getJsonInput();
    
    $pdo = getDB();
    
    // Verificar propriedade
    $stmt = $pdo->prepare("
        SELECT t.id FROM tasks t
        JOIN todo_lists l ON t.list_id = l.id
        WHERE t.id = ? AND l.user_id = ? AND t.deleted_at IS NULL
    ");
    $stmt->execute([$taskId, $userId]);
    
    if (!$stmt->fetch()) {
        jsonResponse(false, 'Tarefa não encontrada', null, 404);
    }
    
    $updates = [];
    $params = [];
    
    if (isset($data['title'])) {
        $updates[] = "title = ?";
        $params[] = trim($data['title']);
    }
    
    if (isset($data['description'])) {
        $updates[] = "description = ?";
        $params[] = trim($data['description']);
    }
    
    if (isset($data['completed'])) {
        $updates[] = "completed = ?";
        $params[] = $data['completed'] ? 1 : 0;
    }
    
    if (array_key_exists('reminder', $data)) {
        $updates[] = "reminder = ?";
        $params[] = $data['reminder'];
    }
    
    if (empty($updates)) {
        jsonResponse(false, 'Nenhum campo para atualizar', null, 400);
    }
    
    $updates[] = "updated_at = NOW()";
    $params[] = $taskId;
    
    $sql = "UPDATE tasks SET " . implode(', ', $updates) . " WHERE id = ?";
    $stmt = $pdo->prepare($sql);
    $stmt->execute($params);
    
    // Log de sincronização
    logSync($userId, 'task', $taskId, 'update');
    
    // Retornar tarefa atualizada
    getTask($userId, $taskId);
}

function deleteTask(string $userId, string $taskId): void {
    $pdo = getDB();
    
    // Verificar propriedade
    $stmt = $pdo->prepare("
        SELECT t.id FROM tasks t
        JOIN todo_lists l ON t.list_id = l.id
        WHERE t.id = ? AND l.user_id = ?
    ");
    $stmt->execute([$taskId, $userId]);
    
    if (!$stmt->fetch()) {
        jsonResponse(false, 'Tarefa não encontrada', null, 404);
    }
    
    // Soft delete
    $stmt = $pdo->prepare("UPDATE tasks SET deleted_at = NOW() WHERE id = ?");
    $stmt->execute([$taskId]);
    
    // Log de sincronização
    logSync($userId, 'task', $taskId, 'delete');
    
    jsonResponse(true, 'Tarefa deletada com sucesso');
}

function generateUUID(): string {
    return sprintf(
        '%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
        mt_rand(0, 0xffff), mt_rand(0, 0xffff),
        mt_rand(0, 0xffff),
        mt_rand(0, 0x0fff) | 0x4000,
        mt_rand(0, 0x3fff) | 0x8000,
        mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)
    );
}

function logSync(string $userId, string $entityType, string $entityId, string $action): void {
    $pdo = getDB();
    $stmt = $pdo->prepare("
        INSERT INTO sync_log (user_id, entity_type, entity_id, action) 
        VALUES (?, ?, ?, ?)
    ");
    $stmt->execute([$userId, $entityType, $entityId, $action]);
}

File: todo-api-rust/src/handlers/tasks.rs
use actix_web::{web, HttpRequest, HttpResponse};
use chrono::{DateTime, Utc};
use sqlx::MySqlPool;

use crate::config::Config;
use crate::errors::{ApiError, ApiResponse};
use crate::middleware::jwt::get_auth_user;
use crate::models::*;

// Helper: Generate UUID
fn generate_uuid() -> String {
    uuid::Uuid::new_v4().to_string()
}

// Helper: Log sync
async fn log_sync(pool: &MySqlPool, user_id: &str, entity_type: &str, entity_id: &str, action: &str) {
    let _ = sqlx::query(
        "INSERT INTO sync_log (user_id, entity_type, entity_id, action) VALUES (?, ?, ?, ?)"
    )
    .bind(user_id)
    .bind(entity_type)
    .bind(entity_id)
    .bind(action)
    .execute(pool)
    .await;
}

// GET /tasks
pub async fn get_all_tasks(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;

    let tasks: Vec<Task> = sqlx::query_as(
        r#"
        SELECT t.id, t.list_id, t.title, t.description, t.completed, t.reminder, t.created_at, t.updated_at, t.deleted_at
        FROM tasks t
        JOIN todo_lists l ON t.list_id = l.id
        WHERE l.user_id = ? AND t.deleted_at IS NULL AND l.deleted_at IS NULL
        ORDER BY t.completed ASC, t.created_at DESC
        "#
    )
    .bind(&claims.user_id)
    .fetch_all(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    let tasks: Vec<TaskResponse> = tasks.into_iter().map(|t| t.into()).collect();

    Ok(HttpResponse::Ok().json(ApiResponse::success("Tarefas carregadas", tasks)))
}

// GET /lists/{id}/tasks
pub async fn get_tasks_by_list(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
    path: web::Path<String>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;
    let list_id = path.into_inner();

    // Check ownership
    let list_exists: Option<(String,)> = sqlx::query_as(
        "SELECT id FROM todo_lists WHERE id = ? AND user_id = ?"
    )
    .bind(&list_id)
    .bind(&claims.user_id)
    .fetch_optional(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    if list_exists.is_none() {
        return Err(ApiError::not_found("Lista não encontrada"));
    }

    let tasks: Vec<Task> = sqlx::query_as(
        r#"
        SELECT id, list_id, title, description, completed, reminder, created_at, updated_at, deleted_at
        FROM tasks 
        WHERE list_id = ? AND deleted_at IS NULL
        ORDER BY completed ASC, created_at DESC
        "#
    )
    .bind(&list_id)
    .fetch_all(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    let tasks: Vec<TaskResponse> = tasks.into_iter().map(|t| t.into()).collect();

    Ok(HttpResponse::Ok().json(ApiResponse::success("Tarefas carregadas", tasks)))
}

// GET /tasks/{id}
pub async fn get_task(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
    path: web::Path<String>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;
    let task_id = path.into_inner();

    let task: Option<Task> = sqlx::query_as(
        r#"
        SELECT t.id, t.list_id, t.title, t.description, t.completed, t.reminder, t.created_at, t.updated_at, t.deleted_at
        FROM tasks t
        JOIN todo_lists l ON t.list_id = l.id
        WHERE t.id = ? AND l.user_id = ? AND t.deleted_at IS NULL
        "#
    )
    .bind(&task_id)
    .bind(&claims.user_id)
    .fetch_optional(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    let task = task.ok_or_else(|| ApiError::not_found("Tarefa não encontrada"))?;

    Ok(HttpResponse::Ok().json(ApiResponse::success("Tarefa encontrada", TaskResponse::from(task))))
}

// POST /tasks
pub async fn create_task(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
    body: web::Json<CreateTaskRequest>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;

    let id = body.id.clone().unwrap_or_else(generate_uuid);
    let list_id = &body.list_id;
    let title = body.title.trim();
    let description = body.description.clone().unwrap_or_default();
    let completed = body.completed.unwrap_or(false);
    let reminder: Option<DateTime<Utc>> = body.reminder.as_ref().and_then(|r| r.parse().ok());
    let created_at = body.created_at.clone().unwrap_or_else(|| Utc::now().to_rfc3339());

    if list_id.is_empty() || title.is_empty() {
        return Err(ApiError::bad_request("Lista e título são obrigatórios"));
    }

    // Check list ownership
    let list_exists: Option<(String,)> = sqlx::query_as(
        "SELECT id FROM todo_lists WHERE id = ? AND user_id = ? AND deleted_at IS NULL"
    )
    .bind(list_id)
    .bind(&claims.user_id)
    .fetch_optional(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    if list_exists.is_none() {
        return Err(ApiError::not_found("Lista não encontrada"));
    }

    // Check if exists (for sync)
    let existing: Option<(String,)> = sqlx::query_as(
        "SELECT id FROM tasks WHERE id = ?"
    )
    .bind(&id)
    .fetch_optional(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    if existing.is_some() {
        // Update if exists
        sqlx::query(
            r#"
            UPDATE tasks 
            SET title = ?, description = ?, completed = ?, reminder = ?, updated_at = NOW(), deleted_at = NULL
            WHERE id = ?
            "#
        )
        .bind(title)
        .bind(&description)
        .bind(completed)
        .bind(&reminder)
        .bind(&id)
        .execute(pool.get_ref())
        .await
        .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;
    } else {
        // Create new
        sqlx::query(
            r#"
            INSERT INTO tasks (id, list_id, title, description, completed, reminder, created_at) 
            VALUES (?, ?, ?, ?, ?, ?, ?)
            "#
        )
        .bind(&id)
        .bind(list_id)
        .bind(title)
        .bind(&description)
        .bind(completed)
        .bind(&reminder)
        .bind(&created_at)
        .execute(pool.get_ref())
        .await
        .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;
    }

    log_sync(pool.get_ref(), &claims.user_id, "task", &id, "create").await;

    Ok(HttpResponse::Created().json(ApiResponse::success(
        "Tarefa criada com sucesso",
        TaskResponse {
            id,
            list_id: list_id.clone(),
            title: title.to_string(),
            description: if description.is_empty() { None } else { Some(description) },
            completed,
            reminder,
            created_at: Utc::now(),
            updated_at: Utc::now(),
        },
    )))
}

// PUT /tasks/{id}
pub async fn update_task(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
    path: web::Path<String>,
    body: web::Json<UpdateTaskRequest>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;
    let task_id = path.into_inner();

    // Check ownership
    let existing: Option<Task> = sqlx::query_as(
        r#"
        SELECT t.* FROM tasks t
        JOIN todo_lists l ON t.list_id = l.id
        WHERE t.id = ? AND l.user_id = ? AND t.deleted_at IS NULL
        "#
    )
    .bind(&task_id)
    .bind(&claims.user_id)
    .fetch_optional(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    let mut task = existing.ok_or_else(|| ApiError::not_found("Tarefa não encontrada"))?;

    // Update fields
    if let Some(title) = &body.title {
        task.title = title.trim().to_string();
    }

    if let Some(description) = &body.description {
        task.description = if description.is_empty() { None } else { Some(description.clone()) };
    }

    if let Some(completed) = body.completed {
        task.completed = completed;
    }

    if let Some(reminder) = &body.reminder {
        task.reminder = if reminder.is_empty() { None } else { reminder.parse().ok() };
    }

    sqlx::query(
        r#"
        UPDATE tasks 
        SET title = ?, description = ?, completed = ?, reminder = ?, updated_at = NOW()
        WHERE id = ?
        "#
    )
    .bind(&task.title)
    .bind(&task.description)
    .bind(task.completed)
    .bind(&task.reminder)
    .bind(&task_id)
    .execute(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    log_sync(pool.get_ref(), &claims.user_id, "task", &task_id, "update").await;

    // Return updated task
    let task: Task = sqlx::query_as(
        "SELECT * FROM tasks WHERE id = ?"
    )
    .bind(&task_id)
    .fetch_one(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    Ok(HttpResponse::Ok().json(ApiResponse::success("Tarefa atualizada", TaskResponse::from(task))))
}

// DELETE /tasks/{id}
pub async fn delete_task(
    req: HttpRequest,
    pool: web::Data<MySqlPool>,
    config: web::Data<Config>,
    path: web::Path<String>,
) -> Result<HttpResponse, ApiError> {
    let claims = get_auth_user(&req, &config)?;
    let task_id = path.into_inner();

    // Soft delete
    sqlx::query(
        r#"
        UPDATE tasks t
        JOIN todo_lists l ON t.list_id = l.id
        SET t.deleted_at = NOW() 
        WHERE t.id = ? AND l.user_id = ?
        "#
    )
    .bind(&task_id)
    .bind(&claims.user_id)
    .execute(pool.get_ref())
    .await
    .map_err(|e| ApiError::internal(format!("Database error: {}", e)))?;

    log_sync(pool.get_ref(), &claims.user_id, "task", &task_id, "delete").await;

    Ok(HttpResponse::Ok().json(ApiResponse::<()>::success_no_data("Tarefa deletada com sucesso")))
}
